### PR TITLE
Add JsonKey.includeFromJson/includeToJson - allow explicit control of serialization

### PR DESCRIPTION
**Pull request by kevmoo**
...and deprecate `JsonKey.ignore`

Fixes https://github.com/google/json_serializable.dart/issues/24
Fixes https://github.com/google/json_serializable.dart/issues/274
Fixes https://github.com/google/json_serializable.dart/issues/537
Fixes https://github.com/google/json_serializable.dart/issues/569
Fixes https://github.com/google/json_serializable.dart/issues/797
Fixes https://github.com/google/json_serializable.dart/issues/1100
Fixes https://github.com/google/json_serializable.dart/issues/1244

### CODE REVIEW

**Comment by ImadBouirmane** [comment]
Useful, I will use it, can I have a simplify description about this new improvement on Json abilities in our codes !

**Comment by kevmoo** [comment]
@ValentinVignal great question! It comes down to terseness. I believe it's easier to document and explain...at least a bit. I realize it's the same state-space. Happy to discuss further, though.

File: json_serializable/lib/src/field_helpers.dart [hunk 1]
```diff
@@ -69,10 +69,10 @@ class _FieldSet implements Comparable<_FieldSet> {
   }
 }
 
-/// Returns a [Set] of all instance [FieldElement] items for [element] and
+/// Returns a [List] of all instance [FieldElement] items for [element] and
 /// super classes, sorted first by their location in the inheritance hierarchy
 /// (super first) and then by their location in the source file.
-Iterable<FieldElement> createSortedFieldSet(ClassElement element) {
+List<FieldElement> createSortedFieldSet(ClassElement element) {
```
Comment (kevmoo) [inline json_serializable/lib/src/field_helpers.dart:75]: Just a small implementation tweak...

File: json_serializable/lib/src/generator_helper.dart [hunk 2]
```diff
@@ -85,28 +86,44 @@ class GeneratorHelper extends HelperCore with EncodeHelper, DecodeHelper {
       final createResult = createFactory(accessibleFields, unavailableReasons);
       yield createResult.output;
 
-      accessibleFieldSet = accessibleFields.entries
+      final fieldsToUse = accessibleFields.entries
```
Comment (kevmoo) [inline json_serializable/lib/src/generator_helper.dart:89]: Needs to be something sortable...

File: json_serializable/lib/src/generator_helper.dart [hunk 3]
```diff
@@ -85,28 +86,44 @@ class GeneratorHelper extends HelperCore with EncodeHelper, DecodeHelper {
...
           .where((e) => createResult.usedFields.contains(e.key))
           .map((e) => e.value)
-          .toSet();
+          .toList();
+
+      for (var candidate in sortedFields.where((element) =>
+          jsonKeyFor(element).explicitYesToJson &&
+          !fieldsToUse.contains(element))) {
+        fieldsToUse.add(candidate);
```
Comment (kevmoo) [inline json_serializable/lib/src/generator_helper.dart:97]: Need to add candidates BACK even if they are not used in the factory if they are forced to be used for toJSON

File: json_serializable/lib/src/generator_helper.dart [hunk 4]
```diff
@@ -85,28 +86,44 @@ class GeneratorHelper extends HelperCore with EncodeHelper, DecodeHelper {
...
+      }
+
+      fieldsToUse.sort(
+          (a, b) => sortedFields.indexOf(a).compareTo(sortedFields.indexOf(b)));
```
Comment (kevmoo) [inline json_serializable/lib/src/generator_helper.dart:101]: Need the fields to maintain the original source ordering

File: json_serializable/lib/src/generator_helper.dart (hunk 3 above)
Comment (jakemac53) [reply json_serializable/lib/src/generator_helper.dart:97]: make these comments actual comments?

Comment (kevmoo) [reply json_serializable/lib/src/generator_helper.dart:99]: done!

**Comment by feduke-nukem** [comment]
I wonder, how I am able to use these fixes? @kevmoo

**Comment by feduke-nukem** [comment]
> @feduke-nukem – that's because I haven't published it yet! Ha! Let me see if I can do that today

That's will be really nice 😁

**Comment by kevmoo** [comment]
> > @feduke-nukem – that's because I haven't published it yet! Ha! Let me see if I can do that today
> 
> That will be really nice 😁

Published!

**Comment by feduke-nukem** [comment]
> > > @feduke-nukem – that's because I haven't published it yet! Ha! Let me see if I can do that today
> > 
> > That will be really nice 😁
> 
> Published!

Thanks!

File: json_serializable/lib/src/generator_helper.dart [hunk 5]
```diff
@@ -123,3 +143,13 @@ class GeneratorHelper extends HelperCore with EncodeHelper, DecodeHelper {
...
 }
+
+extension on KeyConfig {
+  bool get explicitYesFromJson => includeFromJson == true;
+
+  bool get explicitNoFromJson => includeFromJson == false;
+
+  bool get explicitYesToJson => includeToJson == true;
+
+  bool get explicitNoToJson => includeFromJson == false;
```
Comment (mernen) [inline json_serializable/lib/src/generator_helper.dart:154]: Shouldn't this be `includeToJson == false`?

**Pull request merged**